File: composetrack/src/main/java/com/hym/composetrack/LayoutNodeInfo.java
package com.hym.composetrack;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.compose.runtime.ComposeNodeLifecycleCallback;
import androidx.compose.runtime.Recomposer;
import androidx.compose.runtime.collection.MutableVector;
import androidx.compose.ui.Modifier;
import androidx.compose.ui.geometry.Rect;
import androidx.compose.ui.layout.LayoutCoordinates;
import androidx.compose.ui.layout.LayoutCoordinatesKt;
import androidx.compose.ui.layout.LayoutInfo;
import androidx.compose.ui.layout.MeasurePolicy;
import androidx.compose.ui.layout.Remeasurement;
import androidx.compose.ui.node.LayoutNode;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import kotlin.Unit;

/**
 * @author hehua2008
 * @date 2021/12/1
 */
public class LayoutNodeInfo {
    private static final String TAG = "LayoutNodeInfo";

    @NonNull
    private final WeakReference<LayoutNode> mNode;

    public LayoutNodeInfo(@NonNull LayoutInfo node) {
        mNode = new WeakReference((LayoutNode) node);
    }

    @Nullable
    public ComposeNodeLifecycleCallback asComposeNodeLifecycleCallback() {
        return mNode.get();
    }

    @Nullable
    public Remeasurement asRemeasurement() {
        return mNode.get();
    }

    @Nullable
    public LayoutInfo asLayoutInfo() {
        return mNode.get();
    }

    /**
     * @return Coordinates of just the contents of the LayoutNode, after being affected by all
     * modifiers.
     */
    @Nullable
    public LayoutCoordinates getCoordinates() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return node.getCoordinates();
    }

    /**
     * @return The boundaries of this layout relative to the window's origin.
     */
    @Nullable
    public Rect getBoundsInWindow() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutCoordinatesKt.boundsInWindow(node.getCoordinates());
    }

    /**
     * @return Blocks that define the measurement and intrinsic measurement of the layout.
     */
    @Nullable
    public MeasurePolicy getMeasurePolicy() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return node.getMeasurePolicy();
    }

    /**
     * @return The Modifier currently applied to this node.
     */
    @Nullable
    public Modifier getModifier() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return node.getModifier();
    }

    /**
     * @return Apply the specified Modifier to this node.
     */
    public void setModifier(@NonNull Modifier modifier) {
        LayoutNode node = mNode.get();
        if (node == null) return;
        node.setModifier(modifier);
    }

    /**
     * @return The view system Owner. This null until attach is called.
     */
    @Nullable
    public Object getOwner() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return node.getOwner$ui_release();
    }

    /**
     * @return The view system Owner. This null until attach is called.
     */
    @Nullable
    public View getOwnerView() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutNodeHelper.getOwnerView(node);
    }

    /**
     * @return The parent [Recomposer] for this point in the view hierarchy, or `null` if none can
     * be found.
     */
    @Nullable
    public Recomposer getRecomposer() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutNodeHelper.getRecomposer(node);
    }

    /**
     * @return The tree depth of the LayoutNode. This is valid only when it is attached to a
     * hierarchy.
     */
    public int getDepth() {
        LayoutNode node = mNode.get();
        if (node == null) return 0;
        return node.getDepth$ui_release();
    }

    /**
     * @return The size of children.
     */
    @Nullable
    public int getChildrenSize() {
        LayoutNode node = mNode.get();
        if (node == null) return 0;
        return node.get_children$ui_release().getSize();
    }

    /**
     * @return The children of this LayoutNode.
     */
    @Nullable
    public List<LayoutNodeInfo> getChildren() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        MutableVector<LayoutNode> nodeChildren = node.get_children$ui_release();
        List<LayoutNodeInfo> nodeInfoChildren = new ArrayList<>(nodeChildren.getSize());
        nodeChildren.forEach(layoutNode -> {
            nodeInfoChildren.add(new LayoutNodeInfo(layoutNode));
            return Unit.INSTANCE;
        });
        return nodeInfoChildren;
    }

    /**
     * @return The parent node in the LayoutNode hierarchy, skipping over virtual nodes.
     */
    @Nullable
    public LayoutNodeInfo getParent() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        LayoutNode parentNode = node.getParent$ui_release();
        if (parentNode == null) return null;
        return new LayoutNodeInfo(parentNode);
    }

    /**
     * @return All ancestor nodes in the LayoutNode hierarchy, from root LayoutNode to direct parent
     * LayoutNode.
     */
    @Nullable
    public List<LayoutNodeInfo> getAncestors() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        LayoutNodeInfo parent = getParent();
        if (parent == null) return null;
        List<LayoutNodeInfo> ancestors = new ArrayList<>(getDepth());
        while (parent != null) {
            ancestors.add(0, parent);
            parent = parent.getParent();
        }
        return ancestors;
    }

    /**
     * @return The all modifiers of the specified LayoutNode.
     */
    @Nullable
    public List<Modifier> getModifiers() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutNodeHelper.getModifiers(node);
    }

    /**
     * @return The tag associated to a composable with the Modifier.trackId modifier.
     */
    @Nullable
    public TrackId getTrackId() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutNodeHelper.getTrackId(node);
    }

    /**
     * @return TrackId path from root LayoutNode to this LayoutNode.
     */
    @Nullable
    public String getTrackIdPath() {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        return LayoutNodeHelper.getTrackIdPath(node);
    }

    /**
     * @return The LayoutNodeInfo with the specified modifier in the LayoutNode hierarchy,
     * or null` if none can be found.
     */
    @Nullable
    public LayoutNodeInfo findLayoutNodeInfo(@NonNull Modifier modifier)
            throws IllegalStateException {
        LayoutNode node = mNode.get();
        if (node == null) return null;
        final LayoutNode foundNode = LayoutNodeHelper.findLayoutNode(node, modifier);
        if (foundNode == null) return null;
        return new LayoutNodeInfo(foundNode);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (!(obj instanceof LayoutNodeInfo)) return false;
        LayoutNode node = mNode.get();
        LayoutNode anotherNode = ((LayoutNodeInfo) obj).mNode.get();
        return node != null && node.equals(anotherNode);
    }

    @NonNull
    @Override
    public String toString() {
        LayoutNode node = mNode.get();
        if (node == null) return "null";
        return LayoutNodeHelper.layoutNodeToString(node);
    }

    @NonNull
    public String toTreeString() {
        LayoutNode node = mNode.get();
        if (node == null) return "null";
        return LayoutNodeHelper.toTreeString(node);
    }

    public static void travelLayoutNodeTree(@NonNull LayoutNodeInfo nodeInfo,
                                            @NonNull LayoutNodeInfoConsumer consumer) {
        consumer.consume(nodeInfo);
        for (LayoutNodeInfo child : nodeInfo.getChildren()) {
            travelLayoutNodeTree(child, consumer);
        }
    }

    @Nullable
    public static LayoutNodeInfo findInLayoutNodeTree(@NonNull LayoutNodeInfo nodeInfo,
                                                      @NonNull LayoutNodeInfoPredicate predicate) {
        if (predicate.test(nodeInfo)) return nodeInfo;
        for (LayoutNodeInfo child : nodeInfo.getChildren()) {
            LayoutNodeInfo childFound = findInLayoutNodeTree(child, predicate);
            if (childFound != null) return childFound;
        }
        return null;
    }

    public interface LayoutNodeInfoConsumer {
        void consume(@NonNull LayoutNodeInfo nodeInfo);
    }

    public interface LayoutNodeInfoPredicate {
        boolean test(@NonNull LayoutNodeInfo nodeInfo);
    }
}

File: composetrack/src/main/java/com/hym/composetrack/LayoutNodeInfo.kt
package com.hym.composetrack

import android.view.View
import androidx.compose.runtime.ComposeNodeLifecycleCallback
import androidx.compose.runtime.Recomposer
import androidx.compose.runtime.collection.MutableVector
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.layout.LayoutCoordinates
import androidx.compose.ui.layout.LayoutInfo
import androidx.compose.ui.layout.MeasurePolicy
import androidx.compose.ui.layout.Remeasurement
import androidx.compose.ui.layout.boundsInWindow
import androidx.compose.ui.node.LayoutNode
import java.lang.ref.WeakReference

/**
 * @author hehua2008
 * @date 2021/12/1
 */
class LayoutNodeInfo(node: LayoutInfo) {
    private val mNode: WeakReference<LayoutNode?>

    init {
        mNode = WeakReference(node as LayoutNode)
    }

    fun asComposeNodeLifecycleCallback(): ComposeNodeLifecycleCallback? {
        return mNode.get()
    }

    fun asRemeasurement(): Remeasurement? {
        return mNode.get()
    }

    fun asLayoutInfo(): LayoutInfo? {
        return mNode.get()
    }

    val coordinates: LayoutCoordinates?
        /**
         * @return Coordinates of just the contents of the LayoutNode, after being affected by all
         * modifiers.
         */
        get() {
            val node = mNode.get() ?: return null
            return node.coordinates
        }

    val boundsInWindow: Rect?
        /**
         * @return The boundaries of this layout relative to the window's origin.
         */
        get() {
            val node = mNode.get() ?: return null
            return node.coordinates.boundsInWindow()
        }

    val measurePolicy: MeasurePolicy?
        /**
         * @return Blocks that define the measurement and intrinsic measurement of the layout.
         */
        get() {
            val node = mNode.get() ?: return null
            return node.measurePolicy
        }

    var modifier: Modifier?
        /**
         * @return The Modifier currently applied to this node.
         */
        get() {
            val node = mNode.get() ?: return null
            return node.modifier
        }
        /**
         *  Apply the specified Modifier to this node.
         */
        set(value) {
            value ?: return
            val node = mNode.get() ?: return
            node.setModifier(value)
        }

    val owner: Any?
        /**
         * @return The view system Owner. This null until attach is called.
         */
        get() {
            val node = mNode.get() ?: return null
            return node.`owner$ui_release`
        }

    val ownerView: View?
        /**
         * @return The view system Owner. This null until attach is called.
         */
        get() {
            val node = mNode.get() ?: return null
            return LayoutNodeHelper.getOwnerView(node)
        }

    val recomposer: Recomposer?
        /**
         * @return The parent [Recomposer] for this point in the view hierarchy, or `null` if none can
         * be found.
         */
        get() {
            val node = mNode.get() ?: return null
            return LayoutNodeHelper.getRecomposer(node)
        }

    val depth: Int
        /**
         * @return The tree depth of the LayoutNode. This is valid only when it is attached to a
         * hierarchy.
         */
        get() {
            val node = mNode.get() ?: return 0
            return node.`depth$ui_release`
        }

    val childrenSize: Int
        /**
         * @return The size of children.
         */
        get() {
            val node = mNode.get() ?: return 0
            return node.`_children$ui_release`.size
        }

    val children: List<LayoutNodeInfo>?
        /**
         * @return The children of this LayoutNode.
         */
        get() {
            val node = mNode.get() ?: return null
            val nodeChildren: MutableVector<LayoutNode> = node.`_children$ui_release`
            val nodeInfoChildren: MutableList<LayoutNodeInfo> = ArrayList(nodeChildren.size)
            nodeChildren.forEach { layoutNode: LayoutNode ->
                nodeInfoChildren.add(LayoutNodeInfo(layoutNode))
            }
            return nodeInfoChildren
        }

    val parent: LayoutNodeInfo?
        /**
         * @return The parent node in the LayoutNode hierarchy, skipping over virtual nodes.
         */
        get() {
            val node = mNode.get() ?: return null
            val parentNode = node.`parent$ui_release` ?: return null
            return LayoutNodeInfo(parentNode)
        }

    val ancestors: List<LayoutNodeInfo>?
        /**
         * @return All ancestor nodes in the LayoutNode hierarchy, from root LayoutNode to direct parent
         * LayoutNode.
         */
        get() {
            val node = mNode.get() ?: return null
            var parent: LayoutNodeInfo? = parent ?: return null
            val ancestors: MutableList<LayoutNodeInfo> = ArrayList(depth)
            while (parent != null) {
                ancestors.add(0, parent)
                parent = parent.parent
            }
            return ancestors
        }

    val modifiers: List<Modifier>?
        /**
         * @return The all modifiers of the specified LayoutNode.
         */
        get() {
            val node = mNode.get() ?: return null
            return LayoutNodeHelper.getModifiers(node)
        }

    val trackId: TrackId?
        /**
         * @return The tag associated to a composable with the Modifier.trackId modifier.
         */
        get() {
            val node = mNode.get() ?: return null
            return LayoutNodeHelper.getTrackId(node)
        }

    val trackIdPath: String?
        /**
         * @return TrackId path from root LayoutNode to this LayoutNode.
         */
        get() {
            val node = mNode.get() ?: return null
            return LayoutNodeHelper.getTrackIdPath(node)
        }

    /**
     * @return The LayoutNodeInfo with the specified modifier in the LayoutNode hierarchy,
     * or null` if none can be found.
     */
    @Throws(IllegalStateException::class)
    fun findLayoutNodeInfo(modifier: Modifier): LayoutNodeInfo? {
        val node = mNode.get() ?: return null
        val foundNode = LayoutNodeHelper.findLayoutNode(node, modifier) ?: return null
        return LayoutNodeInfo(foundNode)
    }

    override fun hashCode(): Int {
        return super.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other !is LayoutNodeInfo) return false
        val node = mNode.get()
        val anotherNode = other.mNode.get()
        return node != null && node == anotherNode
    }

    override fun toString(): String {
        val node = mNode.get() ?: return "null"
        return LayoutNodeHelper.layoutNodeToString(node)
    }

    fun toTreeString(): String {
        val node = mNode.get() ?: return "null"
        return LayoutNodeHelper.toTreeString(node)
    }

    interface LayoutNodeInfoConsumer {
        fun consume(nodeInfo: LayoutNodeInfo)
    }

    interface LayoutNodeInfoPredicate {
        fun test(nodeInfo: LayoutNodeInfo): Boolean
    }

    companion object {
        private const val TAG = "LayoutNodeInfo"

        fun traverseLayoutNodeTree(
            nodeInfo: LayoutNodeInfo,
            consumer: LayoutNodeInfoConsumer
        ) {
            consumer.consume(nodeInfo)
            val children = nodeInfo.children ?: return
            for (child in children) {
                traverseLayoutNodeTree(child, consumer)
            }
        }

        fun findInLayoutNodeTree(
            nodeInfo: LayoutNodeInfo,
            predicate: LayoutNodeInfoPredicate
        ): LayoutNodeInfo? {
            if (predicate.test(nodeInfo)) return nodeInfo
            val children = nodeInfo.children ?: return null
            for (child in children) {
                val childFound = findInLayoutNodeTree(child, predicate)
                if (childFound != null) return childFound
            }
            return null
        }
    }
}
